### PR TITLE
sdio: add `sdh` peripherals and modify `sdh_demo`

### PR DESCRIPTION
**Pull request by NanaHigh**
Also fix some code issues.
Tested on M1s Dock, is works fine.
Ok

### CODE REVIEW

File: bouffalo-hal/src/sdio.rs [hunk 1]
```diff
@@ -234,6 +240,8 @@ pub enum AutoCMDMode {
     CMD12 = 1,
     /// Auto CMD23 Enable.
     CMD23 = 2,
+    /// Set this bit to zero.
+    ZERO = 0,
```
Comment (luojia65) [inline bouffalo-hal/src/sdio.rs:244]:
```suggestion
    /// Set this bit to zero.
    None = 0,
```
Should we consider using `None` to indicate that no automatic command is enabled?

File: bouffalo-hal/src/sdio.rs [hunk 2]
```diff
@@ -321,22 +330,23 @@ pub struct Command(u16);
 
 /// CMD Type.
 #[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
-pub enum CmdType {
-    Other,
+pub enum SDHCmdType {
```
Comment (luojia65) [inline bouffalo-hal/src/sdio.rs:333]:
```suggestion
pub enum CommandType {
```
I suggest using `CommandType` as the enum name, as the module name sdio already implies it pertains to an SDH command.

File: bouffalo-hal/src/sdio.rs [hunk 3]
```diff
@@ -3510,20 +3533,470 @@ impl TUNINGConfiguration {
...
+        let tmp_val = self.sdh.software_reset.read().reset_all();
+        unsafe {
+            self.sdh.software_reset.write(tmp_val);
+        }
+        while !self.sdh.software_reset.read().is_reset_all_finished() {
+            // Wait for software reset finished.
+            unsafe {
+                asm!("nop");
+            }
+        }
```
Comment (luojia65) [inline bouffalo-hal/src/sdio.rs:3618]:
```suggestion
        // Wait for software reset finished.
        while !self.sdh.software_reset.read().is_reset_all_finished() {
            core::hint::spin_loop()
        }
```

Use `core::hint::spin_loop()` for busy-wait loops. This function provides appropriate hint instructions for each architecture to temporarily reduce the instruction retirement rate, improving energy efficiency.

File: bouffalo-hal/src/sdio.rs [hunk 4]
```diff
@@ -3510,20 +3533,470 @@ impl TUNINGConfiguration {
...
+        tmp_val = tmp_val.set_sdh_clk_sel(0); // GLB_REG_SDH_CLK_SEL.
+        tmp_val = tmp_val.set_sdh_clk_div_len(7); // GLB_REG_SDH_CLK_DIV.
+        unsafe {
+            glb.sdh_config.write(tmp_val);
+        }
+        tmp_val = glb.sdh_config.read();
+        tmp_val = tmp_val.enable_sdh_clk(); // GLB_REG_SDH_CLK_EN.
+        unsafe {
+            glb.sdh_config.write(tmp_val);
+        }
```
Comment (luojia65) [inline bouffalo-hal/src/sdio.rs:3636]: There are multiple writes to the `sdh_config` register. Will the code function correctly if we retain only the last `write`?

File: bouffalo-hal/src/sdio.rs [hunk 5]
```diff
@@ -3510,20 +3533,470 @@ impl TUNINGConfiguration {
...
+        // SDH_Ctrl_Init.
+        let mut tmp_val = self.sdh.clock_control.read();
+        tmp_val = tmp_val.set_sd_clk_freq(0); // SDH_SD_FREQ_SEL_LO.
+        tmp_val = tmp_val.set_sd_clk_freq_upper(0); // SDH_SD_FREQ_SEL_HI.
+        tmp_val = tmp_val.set_clk_gen_mode(ClkGenMode::DividedClk); // SDH_CLK_GEN_SEL.
+        tmp_val = tmp_val.enable_internal_clk(); // SDH_INT_CLK_EN.
+        tmp_val = tmp_val.enable_sd_clk(); // SDH_SD_CLK_EN.
+        unsafe {
+            self.sdh.clock_control.write(tmp_val);
+        }
```
Comment (luojia65) [inline bouffalo-hal/src/sdio.rs:3647]: Prefer `modify` function on `RW` registers to shorten the code size.

**Review by luojia65 — CHANGES_REQUESTED** [review]
The code looks good overall, with only minor style adjustments needed.

**Comment by NanaHigh** [comment]
I have completed the adjustment in latest commit. @luojia65

**Review by luojia65 — APPROVED** [review]
LGTM

**Pull request merged**